scipy/sparse/_index.py

```diff
@@ -293,100 +383,23 @@ def _set_arrayXarray_sparse(self, row, col, x):
         self._set_arrayXarray(row, col, x)
 
 
-def _unpack_index(index) -> tuple[
-    int | slice | npt.NDArray[np.bool_ | np.int_],
-    int | slice | npt.NDArray[np.bool_ | np.int_]
-]:
-    """ Parse index. Always return a tuple of the form (row, col).
-    Valid type for row/col is integer, slice, array of bool, or array of integers.
-    """
-    # Parse any ellipses.
-    index = _check_ellipsis(index)
-
-    # Next, parse the tuple or object
-    if isinstance(index, tuple):
-        if len(index) == 2:
-            row, col = index
-        elif len(index) == 1:
-            row, col = index[0], slice(None)
-        else:
-            raise IndexError('invalid number of indices')
-    else:
-        idx = _compatible_boolean_index(index)
-        if idx is None:
-            row, col = index, slice(None)
-        elif idx.ndim < 2:
-            return idx, slice(None)
-        elif idx.ndim == 2:
-            return idx.nonzero()
-    # Next, check for validity and transform the index as needed.
-    from ._base import issparse
-    if issparse(row) or issparse(col):
-        # Supporting sparse boolean indexing with both row and col does
-        # not work because spmatrix.ndim is always 2.
-        raise IndexError(
-            'Indexing with sparse matrices is not supported '
-            'except boolean indexing where matrix and index '
-            'are equal shapes.')
-    return row, col
-
-
-def _check_ellipsis(index):
-    """Process indices with Ellipsis. Returns modified index."""
-    if index is Ellipsis:
-        return (slice(None), slice(None))
-
-    if not isinstance(index, tuple):
-        return index
-
-    # Find any Ellipsis objects.
-    ellipsis_indices = [i for i, v in enumerate(index) if v is Ellipsis]
-    if not ellipsis_indices:
-        return index
-    if len(ellipsis_indices) > 1:
-        raise IndexError("an index can only have a single ellipsis ('...')")
-
-    # Replace the Ellipsis object with 0, 1, or 2 null-slices as needed.
-    i, = ellipsis_indices
-    num_slices = max(0, 3 - len(index))
-    return index[:i] + (slice(None),) * num_slices + index[i + 1:]
-
-
-def _maybe_bool_ndarray(idx):
-    """Returns a compatible array if elements are boolean.
-    """
-    idx = np.asanyarray(idx)
+def _compatible_boolean_index(idx, desired_ndim):
+    """Check for boolean array or array-like. peek before asarray for array-like"""
+    # assume already an array if attr ndim exists: skip to bottom
+    if not hasattr(idx, 'ndim'):
```

Comment:
This feels a bit risky; but not sure there's any single attribute that would tell us whether we are dealing with a numpy array or any array-like derivative. Maybe `.__array__`, but again not sure what external arrays would implement; we'd perhaps have to check with someone who works on the Array API.

Reply:
I agree that it feels risky -- but it is following the logic of the previous code (the removed line 384 below) which used `ndim` as an indicator of a compatible array type.

I will look for a more appropriate check for compatible array types, but I think that should be a separate PR.
I have added/improved comments explaining that this assumption is being made. Also for better clarity, I added `return None` at the end of this function to be more clear that returning None carries information here .